fix(coverage): Allow baseline coverage per test

nicojs opened the pull request:
Allow coverage result per test to use a `baseline` coverage (coverage which is true for each test) and deviations on that baseline per test. That way the test amount of coverage data (in bytes) can be significantly reduced.

nicojs committed:
fix(coverage): Allow baseline coverage per test

Allow coverage result per test to use a `baseline` coverage (coverage which is true for each test) and deviations on that baseline per test. That way the test amount of coverage data (in bytes) can be significantly reduced.

simondel commented:
So, the baseline contains every statement, while the deviations only contain the statements that are actually hit. And the deviations are the only thing that is sent between processes, right?

nicojs committed:
chore(bithound): Add matched mutant to ignore

Add matched mutant to bithound ignore until bithound supports readonly properties

nicojs commented:
No. All the statements in the file is what we call a `statementMap`. It isn't send cross process, instead we read it in [stryker right after instrumenting the code](https://github.com/stryker-mutator/stryker/blob/master/src/coverage/CoverageInstrumenterStream.ts#L39).

Instead, baseline contains information about every statement hit *before* any test is ran.

Take this js file for example:

```javascript
(function(){ 
  window.MyClass = function(){ 
       this.something = 3;
  }
})(window);
```

If you would load this file into the test runner, the statement at line 0 will be hit before a test is even ran. That coverage is what i call the `baseline` coverage.

The deviations is when a test actually hit the line, so does a `new MyClass()`.

Both the `baseline` and `deviations` are send cross process, because we need both. If a mutant is covered by the `baseline` than it is in fact covered by *every test*, so we'll have to match all tests in that case.

Do you think we have to rename things to make it more clear?

nicojs commented:
As discussed in https://github.com/stryker-mutator/stryker/issues/214

simondel commented:
In your example, the baseline only contains the info about statement 0 right?

nicojs commented:
No, it will contain info about every statement. Most will have `0` as a count. Indeed we could trim down tge amount of data even further by removing all 0's but lets keep it simple for now.

simondel commented:
OK cool!  Ship it!
